File: lib/server_side_dash/aggregator.js
var fs = require('fs')
  , path = require('path')
  , _ = require('underscore')
  , files = {}

var EndDash = module.exports = function(opts) {
  this.templateDir = opts.templateDirPath || "templates/"
  this.aggregatedFilesDir = opts.serveRoot || ""
  this.currentDir = this.templateDir
  this.files = files
  this.loadFiles()
}

Endash.prototype.loadFiles = function() {
  var fileNames = fs.readdir(this.templateDir, this.parseDirectory)
  if(fileNames.length == 0) {
    throw new Error("no templates found in " + this.templateDir)
  }
  this.parseDirectory(fileNames)
}

Endash.prototype.writeFiles = function() {
  var fd = fs.openSync(this.aggregatedFilesDir + "/EndDashTemplates.js", "w")
  _.each(Object.keys(files), function(fileName){
    var bytesWritten = fs.writeSync(fd, files.fileName)
  })
  fs.closeSync(fd)
}

Endash.prototype.parseDirectory = function(err, fileNames) { //Consider sym links and other odd cases

  _.each(fileNames, function(fileName) {
    if(fs.statSync(this.currentDir + fileName).isFile()){
      this.handleFile(fileName)
    } else {
      this.handleDirectory(fileName)
    }
  })

}

Endash.prototype.handleFile = function(fileName) {
  var path = this.pwd + fileName
  if (!this.files.path) {
    this.files.path = fs.readFileSync(path, 'utf8')
  }
}

Endash.prototype.handleDirectory = function(nextDir) {
  var lastPwd = this.pwd
  this.pwd = this.pwd + nextDir
  this.parseDirectory(fs.readdir(this.pwd))
  this.pwd = lastPwd
}

File: lib/server_side_dash/aggregator.rb
class EndDashTemplateAggregator
  @@files = {}
  def aggregate!(opts = {})
    @templateDir = @pwd = opts[:templateDir] || "templates/"
    @aggregatedFilesDir = opts[:serveRoot] || File.dirname(__FILE__)
    @files = @@files
    loadFiles!
    writeFiles!
  end

  def loadFiles!
    currDir = Dir.new(@templateDir)
    raise "not files in #{@templateDir}" unless currDir.entries.length > 0
    parseDirectory!
  end

  def parseDirectory!
    dir = Dir.new(@pwd)
    dir.entries.each do |fileName|
      if fileName != "." && fileName != ".."
        if File.directory?(@pwd + fileName)
          handleDirectory!(fileName)
        else
          handleFile!(fileName)
        end
      end
    end
  end

  def handleFile!(fileName)
    fileStr = ""
    path = "#{@pwd}#{fileName}"
    File.open(path, "r").each_line do |line|
      fileStr = fileStr + line
    end
    @files[path.to_sym] = fileStr
  end

  def handleDirectory!(nextDir)
    oldDir = @pwd
    @pwd = "#{@pwd}#{nextDir}/"
    parseDirectory!
    @pwd = oldDir
  end

  def writeFiles!
    File.open("#{@aggregatedFilesDir}EndDashTemplates.js", 'w'){ |f|
      @files.each do |key, value|
        templateName = key.to_s.gsub(/.*templates\//, "")
        f.write("<script type='EndDash' name='#{templateName}'>\n")
        f.write(value)
        f.write("\n</script>\n")
      end
    }
  end

end

EndDashTemplateAggregator.new.aggregate!({
  templateDir: "/Users/TheOwner/Amicus/end-dash/test/support/templates/",
  serveRoot: "/Users/TheOwner/Amicus/end-dash/test/"
})